### ISSUE TITLE
Generic: presence of callback argument to a factory changes generic parameter inference result

### ISSUE THREAD

**Issue by winhillsen** [Microsoft/TypeScript#24739]
The behavior demonstrated below is stopping me from creating a generic class factory function, whose arguments are callbacks with `this` being the type of the instance the factory's result would produce. Doing so would break correct inference of the generic parameter. I then reduced my issue into its most basic form where no instantiation or class is involved.

**TypeScript Version:**  3.0.0-dev.20180606

**Search Terms:** 
class factory generic, class method this context, arrow function vs normal function difference, generic parameter inference 

**Code**

```ts
interface ITest<T> {}

function factory<T>(method: (ignored: T) => T): ITest<T> {
    return {};
}

const good = factory(() => 42);
const bad = factory(ignored => 42); // !! expect ITest<number> but has ITest<{}>

// With this example, presumably the "normal" function variant implies presence of "this" argument
function factoryWithThis<T>(method: (this: T) => T): ITest<T> {
    return {};
}

const nice = factoryWithThis(() => 42);
const wonky = factoryWithThis(function () { return 42; }); // !! expect ITest<number> but has ITest<{}>
```

**Expected behavior:**
All of `good`, `bad`, `nice` and `wonky` should be of type `ITest<number>`.

**Actual behavior:**
See code comments :point_up:

[**Playground Link**](https://www.typescriptlang.org/play/#src=interface%20ITest%3CT%3E%20%7B%7D%0D%0A%0D%0Afunction%20factory%3CT%3E(method%3A%20(ignored%3A%20T)%20%3D%3E%20T)%3A%20ITest%3CT%3E%20%7B%0D%0A%20%20%20%20return%20%7B%7D%3B%0D%0A%7D%0D%0A%0D%0Aconst%20good%20%3D%20factory(()%20%3D%3E%2042)%3B%0D%0Aconst%20bad%20%3D%20factory(ignored%20%3D%3E%2042)%3B%20%2F%2F%20!!%20expect%20ITest%3Cnumber%3E%20but%20has%20ITest%3C%7B%7D%3E%0D%0A%0D%0A%0D%0Afunction%20factoryWithThis%3CT%3E(method%3A%20(this%3A%20T)%20%3D%3E%20T)%3A%20ITest%3CT%3E%20%7B%0D%0A%20%20%20%20return%20%7B%7D%3B%0D%0A%7D%0D%0A%0D%0Aconst%20nice%20%3D%20factoryWithThis(()%20%3D%3E%2042)%3B%0D%0Aconst%20wonky%20%3D%20factoryWithThis(function%20()%20%7B%20return%2042%3B%20%7D)%3B%20%2F%2F%20!!%20expect%20ITest%3Cnumber%3E%20but%20has%20ITest%3C%7B%7D%3E)

~**Related Issues:**~ (couldn't find any, sorry). If there is a deep reason as to why this is expected behavior I would happily learn from, but it was quite tough to searching for discussions on this.

**Comment by RyanCavanaugh** [comment]
> `const bad = factory(ignored => 42);`

Here's what happens during this resolution:
 * We need to find candidates for `factory`'s `T`
 * There are two candidate locations: the argument, and the return type of the body
   * What's the type of the argument? It's unannotated, so we check for a contextual type
     * There's a contextual type, but it's generic, so it resolves as `{}` (the constraint type)
   * What's the type of the return type? The function returns, but we assume the body of the function actually depends on its parameters, so we can't compute its return type until the parameter types are resolved first
 * Only one candidate survives, and it's `{}`

Basically we can't tell this situation from the case where you wrote:
```ts
const wat = factory(ignored => ignored.moo);
```

You could instead write `factory` as
```ts
function factory<T, U extends T>(method: (ignored: T) => U): ITest<U> {
    return {};
}
```
Or instead declare that you're not looking at `ignored`:
```ts
const bad = factory((ignored: never) => 42);
```

The example with `this` is similar; the arrow function works because we know that the arrow function can't care about the calling `this`.

**Comment by winhillsen** [comment]
I see, yes! Thank you for the explanation.
